app/main/views/suppliers.py

```diff
@@ -512,25 +513,49 @@ def create_new_supplier():
 def duns_number():
     form = DunsNumberForm()
 
-    if request.method == 'POST':
-        if form.validate_on_submit():
-            suppliers = data_api_client.find_suppliers(duns_number=form.duns_number.data)
-            if len(suppliers["suppliers"]) <= 0:
-                session[form.duns_number.name] = form.duns_number.data
-                return redirect(url_for(".company_details"))
+    if request.method == 'GET':
+        if request.args.get('retry', None) == 'true':
+            # When redirected back from the lookup page
+            form.duns_number.errors = ["Enter a different DUNS number"]
+        elif form.duns_number.name in session:
+            # Likely a back click or unexpected navigation, refill form data
+            form.duns_number.data = session[form.duns_number.name]
+
+    if form.validate_on_submit():
+        # Valid duns number
 
+        # Fail on existing account with this duns number first
+        if len(data_api_client.find_suppliers(duns_number=form.duns_number.data)['suppliers']) > 0:
             form.duns_number.errors = ["DUNS number already used"]
+        else:
+            # Check for number in Direct Plus API
+            # If we cannot contact the API then skip this step
+            # If the client doesn't return an organization then there is not a record for that DUNs
+            # Otherwise the number is good, add the 'primaryName' to the session and continue
+            try:
+                organization = direct_plus_client.get_organization_by_duns_number(form.duns_number.data)
```

Comment:
_I_ approve of `organization` with a Z (I would personally use american spelling in all variable names) but note that we seem to use `organisation` throughout the rest of the code.

Reply:
Of course the cheat to avoid the subject would be to just call it `org` everywhere

Reply:
Yeah, I've kept it in link with what it's named in the response from DnB. American throughout would probably not be a bad idea.